engine/access/rpc/backend/connection_factory.go

```diff
@@ -64,10 +68,30 @@ type ConnectionFactoryImpl struct {
 }
 
 type CachedClient struct {
-	ClientConn *grpc.ClientConn
-	Address    string
-	mutex      sync.Mutex
-	timeout    time.Duration
+	ClientConn     *grpc.ClientConn
+	Address        string
+	mutex          sync.Mutex
+	timeout        time.Duration
+	requestCounter *atomic.Int32
+	closeRequested *atomic.Bool
+	done           chan struct{}
+}
+
+func (s *CachedClient) UpdateConnection(conn *grpc.ClientConn) {
+	s.ClientConn = conn
+	s.closeRequested.Store(false)
+	s.requestCounter.Store(0)
+	s.done = make(chan struct{}, 1)
+}
+
+func NewCachedClient(timeout time.Duration, address string) *CachedClient {
+	return &CachedClient{
+		Address:        address,
+		timeout:        timeout,
+		requestCounter: atomic.NewInt32(0),
+		closeRequested: atomic.NewBool(false),
+		done:           make(chan struct{}, 1),
```

Comment:
what do you think about using a wait group instead of a channel and counter? The functionality is basically the same, but `sync.WaitGroup` is thoroughly tested and a well established pattern